File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/pair_with_target_sum_binary_search.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int binary_search(const vector<int> &a, int low, int high, int elem) {
  int mid = -1;

  while (low <= high) {
    mid = low + (high - low) / 2;

    if (a[mid] == elem) {
      return mid;
    } else if (a[mid] < elem) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }

  return -1;
}

// O(nlogn) / O(1)
pair<int, int> pair_with_sum(const vector<int> &a, int s) {
  int lpos = -1, rpos = -1, idx = -1;

  for (int i = 0; i < a.size(); i++) {
    idx = binary_search(a, 0, a.size() - 1, s - a[i]);
    if (idx != -1) {
      lpos = i;
      rpos = idx;
      break;
    }
  }

  return make_pair<>(lpos, rpos);
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, s;
  cin >> tt;

  while (tt--) {
    cin >> n >> s;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    auto [pos1, pos2] = pair_with_sum(a, s);
    cout << pos1 << " " << pos2 << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/PairWithTargetSumBinarySearch.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class PairWithTargetSumBinarySearch {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int s = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        var res = pairWihTargetSum(a, s);
        System.out.printf("%d %d\n", res[0], res[1]);
      }
    }
  }

  // O(nlogn) / O(1)
  public static int[] pairWihTargetSum(int[] a, int s) {
    int[] pos = new int[] {-1, -1};

    for (int i = 0; i < a.length - 1; i++) {
      int idx = binarySearch(a, 0, a.length - 1, s - a[i]);

      if (idx != -1) {
        pos[0] = i;
        pos[1] = idx;
        break;
      }
    }

    return pos;
  }

  public static int binarySearch(int[] a, int low, int high, int elem) {
    while (low <= high) {
      int mid = low + (high - low) / 2;

      if (a[mid] == elem) {
        return mid;
      } else if (a[mid] < elem) {
        low = mid + 1;
      } else {
        high = mid - 1;
      }
    }

    return -1;
  }
}